tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -340,12 +338,83 @@ export class WebGPUBackend extends KernelBackend {
     return res;
   }
 
+  private getAndSaveProgram(
+      key: number, getProgram: () => webgpu_program.WebGPUProgram) {
+    if (!this.programCache.has(key)) {
+      this.programCache.set(key, getProgram());
+    }
+    return this.programCache.get(key);
+  }
+
+  // This hash algorithm came from https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/common/HashUtils.h;l=47;bpv=1;bpt=0
+  private getHash<R extends Rank>(input: Array<boolean|number|ShapeMap[R]|
```

Comment:
Using a hash function in this way is a bit hazardous. This hash function is pretty good, but we can't be totally confident that collisions won't occur. However if we can't find a better solution then it will probably be okay.